File: backend/app/application/use_cases/record_use_cases.py
from typing import List, Optional, Any
from app.domain.entities.record import RecordEntity
from app.domain.events.record_events import record_event_bus, RECORD_CREATED
from app.application.dtos.record_dto import CreateRecordDto, RecordResponseDto

class CreateRecordUseCase:
    def __init__(self, record_repository, search_adapter=None, sns_publisher=None, ses_notifier=None):
        self.record_repository = record_repository
        self.search_adapter = search_adapter
        self.sns_publisher = sns_publisher
        self.ses_notifier = ses_notifier

    async def execute(self, dto: CreateRecordDto) -> RecordResponseDto:
        entity = await self.record_repository.create(title=dto.title, description=dto.description or "")
        wire = entity.to_wire()
        response_dto = RecordResponseDto(**wire)

        # Domain event
        record_event_bus.emit(RECORD_CREATED, response_dto)

        # ElasticSearch indexing
        if self.search_adapter and hasattr(self.search_adapter, "index_record"):
            try:
                await self.search_adapter.index_record(wire)
            except Exception as e:
                print(f"[CreateRecordUseCase] Search indexing warning: {e}")

        # SNS notification
        if self.sns_publisher and hasattr(self.sns_publisher, "publish_record_created"):
            try:
                await self.sns_publisher.publish_record_created(wire)
            except Exception as e:
                print(f"[CreateRecordUseCase] SNS publish warning: {e}")

        # SES notification
        if self.ses_notifier and hasattr(self.ses_notifier, "send_record_created_notification"):
            try:
                await self.ses_notifier.send_record_created_notification(wire)
            except Exception as e:
                print(f"[CreateRecordUseCase] SES notification warning: {e}")

        return response_dto

class GetRecordByIdUseCase:
    def __init__(self, record_repository):
        self.record_repository = record_repository

    async def execute(self, record_id: str) -> Optional[RecordResponseDto]:
        if not record_id:
            raise ValueError("Record ID is required")
        entity = await self.record_repository.find_by_id(record_id)
        if not entity:
            return None
        return RecordResponseDto(**entity.to_wire())

class ListRecordsUseCase:
    def __init__(self, record_repository):
        self.record_repository = record_repository

    async def execute(self, limit: int = 50, skip: int = 0) -> List[RecordResponseDto]:
        entities = await self.record_repository.find_all(limit=limit, skip=skip)
        return [RecordResponseDto(**e.to_wire()) for e in entities]

class SearchRecordsUseCase:
    def __init__(self, search_adapter, record_repository=None):
        self.search_adapter = search_adapter
        self.record_repository = record_repository

    async def execute(self, query: str) -> List[dict]:
        if not query or not query.strip():
            return []

        if self.search_adapter and hasattr(self.search_adapter, "search"):
            try:
                return await self.search_adapter.search(query.strip())
            except Exception as e:
                print(f"[SearchRecordsUseCase] ES search failed: {e}")

        if self.record_repository and hasattr(self.record_repository, "find_by_title_or_description"):
            entities = await self.record_repository.find_by_title_or_description(query.strip())
            return [e.to_wire() for e in entities]

        return []

class ExportFormatUseCase:
    def __init__(self, record_repository):
        self.record_repository = record_repository

    async def execute(self, format_type: str = "json") -> Any:
        entities = await self.record_repository.find_all()
        records = [e.to_wire() for e in entities]

        if format_type == "csv":
            header = "id,title,description,createdAt\n"
            rows = "\n".join(f'"{r["id"]}","{r["title"]}","{r["description"]}","{r["createdAt"]}"' for r in records)
            return header + rows

        if format_type == "xml":
            items = "\n".join(
                f'  <record>\n    <id>{r["id"]}</id>\n    <title>{r["title"]}</title>\n    <description>{r["description"]}</description>\n    <createdAt>{r["createdAt"]}</createdAt>\n  </record>'
                for r in records
            )
            return f"<records>\n{items}\n</records>"

        return records

File: backend/src/application/use-cases/RecordUseCases.ts
import { CreateRecordDto, RecordResponseDto } from '../dtos/Record.dto';
import { recordEventBus, RECORD_EVENTS } from '../../domain/events/recordEvents';

export interface RecordRepositoryPort {
  create(data: { title: string; description: string }): Promise<any>;
  findById(id: string): Promise<any>;
  findAll(options?: { limit?: number; skip?: number }): Promise<any[]>;
  findByTitleOrDescription?(text: string): Promise<any[]>;
}

export interface SearchPort {
  indexRecord?(record: any): Promise<void>;
  search(query: string): Promise<any[]>;
}

export interface SnsPort {
  publishRecordCreated?(record: any): Promise<any>;
}

export interface SesPort {
  sendRecordCreatedNotification?(record: any, recipient?: string): Promise<any>;
}

export class CreateRecordUseCase {
  constructor(
    private readonly dependencies: {
      recordRepository: RecordRepositoryPort;
      searchAdapter?: SearchPort;
      snsPublisher?: SnsPort;
      sesNotifier?: SesPort;
    }
  ) {}

  async execute(input: { title: string; description?: string }): Promise<RecordResponseDto> {
    const dto = new CreateRecordDto(input);
    const savedEntity = await this.dependencies.recordRepository.create(dto);
    const responseDto = new RecordResponseDto(savedEntity);

    recordEventBus.emit(RECORD_EVENTS.RECORD_CREATED, responseDto);

    if (this.dependencies.searchAdapter && typeof this.dependencies.searchAdapter.indexRecord === 'function') {
      try {
        await this.dependencies.searchAdapter.indexRecord(responseDto);
      } catch (err: any) {
        console.warn(`[CreateRecordUseCase] Search indexing warning: ${err.message}`);
      }
    }

    if (this.dependencies.snsPublisher && typeof this.dependencies.snsPublisher.publishRecordCreated === 'function') {
      try {
        await this.dependencies.snsPublisher.publishRecordCreated(responseDto);
      } catch (err: any) {
        console.warn(`[CreateRecordUseCase] SNS publish warning: ${err.message}`);
      }
    }

    if (this.dependencies.sesNotifier && typeof this.dependencies.sesNotifier.sendRecordCreatedNotification === 'function') {
      try {
        await this.dependencies.sesNotifier.sendRecordCreatedNotification(responseDto);
      } catch (err: any) {
        console.warn(`[CreateRecordUseCase] SES notification warning: ${err.message}`);
      }
    }

    return responseDto;
  }
}

export class GetRecordByIdUseCase {
  constructor(private readonly recordRepository: RecordRepositoryPort) {}

  async execute(id: string): Promise<RecordResponseDto | null> {
    if (!id) {
      throw new Error('Record ID is required');
    }
    const recordEntity = await this.recordRepository.findById(id);
    if (!recordEntity) {
      return null;
    }
    return new RecordResponseDto(recordEntity);
  }
}

export class ListRecordsUseCase {
  constructor(private readonly recordRepository: RecordRepositoryPort) {}

  async execute(options: { limit?: number; skip?: number } = {}): Promise<RecordResponseDto[]> {
    const entities = await this.recordRepository.findAll(options);
    return entities.map((entity) => new RecordResponseDto(entity));
  }
}

export class SearchRecordsUseCase {
  constructor(
    private readonly searchAdapter: SearchPort,
    private readonly recordRepository?: RecordRepositoryPort
  ) {}

  async execute(query: string): Promise<any[]> {
    if (!query || typeof query !== 'string' || !query.trim()) {
      return [];
    }

    if (this.searchAdapter && typeof this.searchAdapter.search === 'function') {
      try {
        return await this.searchAdapter.search(query.trim());
      } catch (err: any) {
        console.warn(`[SearchRecordsUseCase] Elasticsearch search failed: ${err.message}. Falling back to DB search.`);
      }
    }

    if (this.recordRepository && typeof this.recordRepository.findByTitleOrDescription === 'function') {
      const entities = await this.recordRepository.findByTitleOrDescription(query.trim());
      return entities.map((e) => new RecordResponseDto(e));
    }

    return [];
  }
}

export class ExportFormatUseCase {
  constructor(private readonly recordRepository: RecordRepositoryPort) {}

  async execute(format: string = 'json'): Promise<any> {
    const records = await this.recordRepository.findAll();
    const dtos = records.map((r) => new RecordResponseDto(r));

    if (format === 'csv') {
      const header = 'id,title,description,createdAt\n';
      const rows = dtos.map((d) => `"${d.id}","${d.title.replace(/"/g, '""')}","${d.description.replace(/"/g, '""')}","${d.createdAt}"`).join('\n');
      return header + rows;
    }

    if (format === 'xml') {
      const items = dtos.map((d) => `  <record>\n    <id>${d.id}</id>\n    <title>${d.title}</title>\n    <description>${d.description}</description>\n    <createdAt>${d.createdAt}</createdAt>\n  </record>`).join('\n');
      return `<records>\n${items}\n</records>`;
    }

    return dtos;
  }
}

export default {
  CreateRecordUseCase,
  GetRecordByIdUseCase,
  ListRecordsUseCase,
  SearchRecordsUseCase,
  ExportFormatUseCase,
};
